compiler/rustc_mir/src/transform/promote_consts.rs

```diff
@@ -545,7 +558,7 @@ impl<'tcx> Validator<'_, 'tcx> {
                     }
 
                     ProjectionElem::Field(..) => {
-                        if self.const_kind.is_none() {
+                        if self.maybe_runtime() {
```

Comment:
I didn't realize this in my first round of comments, but this means that we currently promote union field accesses in constants and statics but not `fn`, which is somewhat weird. Union field accesses are currently unstable in `const fn`, so this doesn't change anything for stable users, but if the goal is to normalize promotion across all const contexts, we need to figure out what to do here eventually.

Reply:
Yes, I mentioned this in https://github.com/rust-lang/const-eval/issues/53.

We cannot accept them in `fn` as union field access are a fallible operation -- they can cause UB. So either we keep them const/static-only, or we never promote them. In my follow-up PR that needs crater anyway, I intend to reject promoting them to see what the fallout of that would be.